File: apps/quickink/ios/QuickInk/Stories/StoryPdfExporter.swift
/*
 * StoryPdfExporter.swift
 *
 * Stories Phase 4 — renders a Story + its items into a US-Letter PDF
 * via `UIGraphicsPDFRenderer`. Page 1 is the cover; subsequent pages
 * carry items. Per-page item count is driven by `StoryItem.Layout`:
 *
 *   - full → 1 item per page (default)
 *   - half → 2 items per page
 *   - grid → 4 items per page
 *
 * Day markers (from `StoryDayMarkers.derive`) render inline above
 * the first item of each new bucket. End-card lands on the final
 * page. The output is byte-deterministic for the same input + clock
 * so callers can dedupe runs.
 *
 * Mirror of Android `StoryPdfExporter.kt`.
 */

import Foundation
import UIKit

enum StoryPdfExporter {

    /// US-Letter portrait at 72 DPI. SwiftUI's PDF context uses
    /// `CGFloat` points — same coordinate space as on-screen views.
    private static let pageRect = CGRect(x: 0, y: 0, width: 612, height: 792)
    private static let margin: CGFloat = 48

    /// Render the story to a temporary file on disk; returns the
    /// file URL or throws if write fails. Caller hands the URL off
    /// to `UIActivityViewController`.
    ///
    /// `previewUris` maps each capture-backed `StoryItem.refId` to
    /// its on-disk `preview_uri`. The exporter loads + embeds the
    /// bitmap when present; missing entries fall back to a cream-box
    /// placeholder. The share sheet builds the map before calling.
    static func export(story: Story, items: [StoryItem], previewUris: [String: String] = [:]) throws -> URL {
        let renderer = UIGraphicsPDFRenderer(bounds: pageRect)
        let dir = FileManager.default.temporaryDirectory
        let safeTitle = sanitizeFileName(story.title)
        let url = dir.appendingPathComponent("\(safeTitle).pdf")

        let markers = StoryDayMarkers.derive(from: items)
        let markerByItemId = Dictionary(uniqueKeysWithValues: markers.map { ($0.precedingItemId, $0) })

        try renderer.writePDF(to: url) { ctx in
            // ---- Cover page ----
            ctx.beginPage()
            drawCover(story: story, items: items)

            // ---- Item pages ----
            let chunks = paginate(items: items)
            for chunk in chunks {
                ctx.beginPage()
                drawChunk(chunk, markerByItemId: markerByItemId, previewUris: previewUris)
            }

            // ---- End-card page ----
            ctx.beginPage()
            drawEndCard(story: story)
        }
        return url
    }

    // MARK: - Cover

    private static func drawCover(story: Story, items: [StoryItem]) {
        // Paper-warm gradient background — uses the same palette as
        // the reader's `StoryCoverColor.gradient`. Drawn manually via
        // CGContext linear gradient.
        guard let ctx = UIGraphicsGetCurrentContext() else { return }
        let colors = coverColors(for: story.coverStyle)
        let gradient = CGGradient(
            colorsSpace: CGColorSpaceCreateDeviceRGB(),
            colors: [colors.0.cgColor, colors.1.cgColor] as CFArray,
            locations: [0, 1]
        )!
        let coverRect = CGRect(x: margin, y: 96, width: pageRect.width - margin * 2, height: 320)
        ctx.saveGState()
        ctx.addPath(UIBezierPath(roundedRect: coverRect, cornerRadius: 16).cgPath)
        ctx.clip()
        ctx.drawLinearGradient(
            gradient,
            start: CGPoint(x: coverRect.minX, y: coverRect.minY),
            end:   CGPoint(x: coverRect.maxX, y: coverRect.maxY),
            options: []
        )
        ctx.restoreGState()

        // Stamp (top-right of cover)
        let stampSource = story.timeRangeStart
            ?? items.first.map { $0.occurredAt ?? $0.createdAt }
            ?? story.createdAt
        let stamp = monthYearStamp(stampSource) ?? ""
        let stampAttrs: [NSAttributedString.Key: Any] = [
            .font: UIFont.italicSystemFont(ofSize: 10),
            .foregroundColor: UIColor.darkGray,
            .kern: 2,
        ]
        (stamp as NSString).draw(
            at: CGPoint(x: coverRect.minX + 16, y: coverRect.minY + 16),
            withAttributes: stampAttrs
        )

        // Title + subtitle (bottom-left of cover)
        let titleFont = UIFont(name: "Lora-Medium", size: 28) ?? UIFont.systemFont(ofSize: 28, weight: .medium)
        let titleAttrs: [NSAttributedString.Key: Any] = [
            .font: titleFont,
            .foregroundColor: UIColor.black,
        ]
        let titleSize = (story.title as NSString).boundingRect(
            with: CGSize(width: coverRect.width - 32, height: .greatestFiniteMagnitude),
            options: [.usesLineFragmentOrigin],
            attributes: titleAttrs,
            context: nil
        ).size
        (story.title as NSString).draw(
            with: CGRect(
                x: coverRect.minX + 16,
                y: coverRect.maxY - 16 - titleSize.height - (story.subtitle?.isEmpty == false ? 26 : 0),
                width: coverRect.width - 32,
                height: titleSize.height
            ),
            options: [.usesLineFragmentOrigin],
            attributes: titleAttrs,
            context: nil
        )

        if let subtitle = story.subtitle, !subtitle.isEmpty {
            let subFont = UIFont(name: "Caveat-Medium", size: 18) ?? UIFont.italicSystemFont(ofSize: 16)
            let subAttrs: [NSAttributedString.Key: Any] = [
                .font: subFont,
                .foregroundColor: UIColor.darkGray,
            ]
            (subtitle as NSString).draw(
                at: CGPoint(x: coverRect.minX + 16, y: coverRect.maxY - 32),
                withAttributes: subAttrs
            )
        }

        // Attribution below the cover
        let attrFont = UIFont.italicSystemFont(ofSize: 11)
        let attribution = "\(items.count) items"
        let attrAttrs: [NSAttributedString.Key: Any] = [
            .font: attrFont, .foregroundColor: UIColor.darkGray,
        ]
        let attribSize = (attribution as NSString).size(withAttributes: attrAttrs)
        (attribution as NSString).draw(
            at: CGPoint(
                x: (pageRect.width - attribSize.width) / 2,
                y: coverRect.maxY + 16
            ),
            withAttributes: attrAttrs
        )
    }

    // MARK: - Items per page

    private struct Chunk {
        let items: [StoryItem]
        let perPage: Int
    }

    private static func paginate(items: [StoryItem]) -> [Chunk] {
        var out: [Chunk] = []
        var i = 0
        while i < items.count {
            let perPage = itemsPerPage(for: items[i].layout)
            let end = min(i + perPage, items.count)
            out.append(Chunk(items: Array(items[i..<end]), perPage: perPage))
            i = end
        }
        return out
    }

    private static func itemsPerPage(for layoutRaw: String) -> Int {
        switch layoutRaw {
        case StoryItem.Layout.half.rawValue: return 2
        case StoryItem.Layout.grid.rawValue: return 4
        default:                             return 1
        }
    }

    // MARK: - Chunk rendering

    private static func drawChunk(_ chunk: Chunk, markerByItemId: [String: StoryDayMarker], previewUris: [String: String]) {
        let contentRect = pageRect.insetBy(dx: margin, dy: margin)
        var yCursor = contentRect.minY

        for (idx, item) in chunk.items.enumerated() {
            if let marker = markerByItemId[item.id] {
                yCursor = drawDayMarker(marker.label, at: yCursor, width: contentRect.width, originX: contentRect.minX)
            }

            let rect: CGRect = {
                switch chunk.perPage {
                case 2:
                    let h = (contentRect.height - 32) / 2
                    return CGRect(x: contentRect.minX, y: yCursor, width: contentRect.width, height: h)
                case 4:
                    let h = (contentRect.height - 32 * 2) / 4
                    return CGRect(x: contentRect.minX, y: yCursor, width: contentRect.width, height: h)
                default:
                    return CGRect(x: contentRect.minX, y: yCursor, width: contentRect.width, height: contentRect.height - (yCursor - contentRect.minY))
                }
            }()

            drawItem(item, in: rect, previewUris: previewUris)
            yCursor = rect.maxY + 24

            if idx == chunk.items.count - 1 { break }
        }
    }

    private static func drawDayMarker(_ label: String, at y: CGFloat, width: CGFloat, originX: CGFloat) -> CGFloat {
        let font = UIFont(name: "Lora-Italic", size: 11) ?? UIFont.italicSystemFont(ofSize: 11)
        let attrs: [NSAttributedString.Key: Any] = [
            .font: font,
            .foregroundColor: UIColor.systemOrange.withAlphaComponent(0.85),
            .kern: 1,
        ]
        let size = (label as NSString).size(withAttributes: attrs)
        (label as NSString).draw(
            at: CGPoint(x: originX, y: y),
            withAttributes: attrs
        )
        return y + size.height + 10
    }

    private static func drawItem(_ item: StoryItem, in rect: CGRect, previewUris: [String: String] = [:]) {
        guard let ctx = UIGraphicsGetCurrentContext() else { return }

        switch item.kind {
        case StoryItem.Kind.textBlock.rawValue:
            let font = UIFont(name: "Lora-Regular", size: 14) ?? UIFont.systemFont(ofSize: 14)
            let para = NSMutableParagraphStyle()
            para.lineSpacing = 4
            let attrs: [NSAttributedString.Key: Any] = [
                .font: font, .foregroundColor: UIColor.black, .paragraphStyle: para,
            ]
            (item.text ?? "").draw(with: rect, options: [.usesLineFragmentOrigin], attributes: attrs, context: nil)

        case StoryItem.Kind.handwrittenNote.rawValue:
            // Left coral rule + Caveat text
            ctx.setFillColor(UIColor.systemOrange.cgColor)
            ctx.fill(CGRect(x: rect.minX, y: rect.minY, width: 2, height: min(rect.height, 60)))
            let font = UIFont(name: "Caveat-Medium", size: 18) ?? UIFont.italicSystemFont(ofSize: 16)
            (item.text ?? "").draw(
                with: rect.insetBy(dx: 12, dy: 0).offsetBy(dx: 6, dy: 0),
                options: [.usesLineFragmentOrigin],
                attributes: [.font: font, .foregroundColor: UIColor.darkGray],
                context: nil
            )

        case StoryItem.Kind.dateDivider.rawValue:
            let font = UIFont(name: "Lora-Italic", size: 12) ?? UIFont.italicSystemFont(ofSize: 12)
            let label = item.text ?? "Date divider"
            (label as NSString).draw(
                at: CGPoint(x: rect.midX - 100, y: rect.minY),
                withAttributes: [.font: font, .foregroundColor: UIColor.systemOrange.withAlphaComponent(0.85), .kern: 1.0]
            )

        case StoryItem.Kind.placePin.rawValue:
            let font = UIFont(name: "Lora-MediumItalic", size: 13) ?? UIFont.italicSystemFont(ofSize: 13)
            ("📍 " + (item.text ?? "")).draw(
                at: CGPoint(x: rect.minX, y: rect.minY),
                withAttributes: [.font: font, .foregroundColor: UIColor.darkGray]
            )

        case StoryItem.Kind.voiceClip.rawValue:
            // Audio doesn't render to PDF; show a transcription-style
            // placeholder card. Phase 5 transcripts (if added) replace
            // the placeholder text.
            ctx.setStrokeColor(UIColor.lightGray.cgColor)
            ctx.setLineWidth(1)
            ctx.addPath(UIBezierPath(roundedRect: rect.insetBy(dx: 0, dy: 8), cornerRadius: 12).cgPath)
            ctx.strokePath()
            let font = UIFont(name: "Lora-Italic", size: 12) ?? UIFont.italicSystemFont(ofSize: 12)
            let label = "🎙  \(item.caption ?? "Voice clip — open the app to listen.")"
            (label as NSString).draw(
                at: CGPoint(x: rect.minX + 12, y: rect.minY + 20),
                withAttributes: [.font: font, .foregroundColor: UIColor.darkGray]
            )

        default:
            // Photo / document — load the capture's preview JPEG and
            // embed it inside `photoRect`. Falls back to a cream box
            // when the preview is missing or unreadable.
            let photoHeight: CGFloat = min(rect.height - 30, 360)
            let photoRect = CGRect(x: rect.minX, y: rect.minY, width: rect.width, height: photoHeight)

            let img: UIImage? = {
                guard let refId = item.refId,
                      let uri   = previewUris[refId] else { return nil }
                return loadImage(at: uri)
            }()

            if let img = img {
                ctx.saveGState()
                ctx.addPath(UIBezierPath(roundedRect: photoRect, cornerRadius: 8).cgPath)
                ctx.clip()
                img.draw(in: aspectFill(image: img.size, into: photoRect))
                ctx.restoreGState()
            } else {
                ctx.setFillColor(UIColor(white: 0.92, alpha: 1).cgColor)
                ctx.addPath(UIBezierPath(roundedRect: photoRect, cornerRadius: 8).cgPath)
                ctx.fillPath()
            }

            if let caption = item.caption, !caption.isEmpty {
                let font = UIFont(name: "Lora-Italic", size: 12) ?? UIFont.italicSystemFont(ofSize: 12)
                (caption as NSString).draw(
                    at: CGPoint(x: rect.minX, y: photoRect.maxY + 6),
                    withAttributes: [.font: font, .foregroundColor: UIColor.darkGray]
                )
            }
        }
    }

    /// Compute the aspect-fill rect for an image of `imageSize`
    /// drawn into `target`. Used by the photo embed so portraits
    /// fill the slot without letterboxing.
    private static func aspectFill(image imageSize: CGSize, into target: CGRect) -> CGRect {
        guard imageSize.width > 0, imageSize.height > 0 else { return target }
        let scale = max(target.width / imageSize.width, target.height / imageSize.height)
        let w = imageSize.width  * scale
        let h = imageSize.height * scale
        return CGRect(
            x: target.midX - w / 2,
            y: target.midY - h / 2,
            width:  w,
            height: h
        )
    }

    /// Load a `UIImage` from a `file://` URI (or a bare path) when
    /// the file exists locally. Nil on miss.
    private static func loadImage(at uri: String) -> UIImage? {
        let path: String?
        if let url = URL(string: uri), url.isFileURL {
            path = url.path
        } else {
            path = uri
        }
        guard let p = path, FileManager.default.fileExists(atPath: p) else { return nil }
        return UIImage(contentsOfFile: p)
    }

    // MARK: - End card

    private static func drawEndCard(story: Story) {
        let endRect = CGRect(x: margin, y: 280, width: pageRect.width - margin * 2, height: 180)
        guard let ctx = UIGraphicsGetCurrentContext() else { return }
        ctx.setStrokeColor(UIColor.lightGray.cgColor)
        ctx.setLineWidth(1)
        ctx.addPath(UIBezierPath(roundedRect: endRect, cornerRadius: 14).cgPath)
        ctx.strokePath()

        let font = UIFont(name: "Lora-Italic", size: 13) ?? UIFont.italicSystemFont(ofSize: 13)
        let endLabel = "— THE END —"
        let endSize = (endLabel as NSString).size(withAttributes: [.font: font, .kern: 1.5])
        (endLabel as NSString).draw(
            at: CGPoint(x: endRect.midX - endSize.width / 2, y: endRect.midY - 12),
            withAttributes: [.font: font, .foregroundColor: UIColor.darkGray, .kern: 1.5]
        )

        let footer = "— Made with QuickInk · scan, jot, find again. —"
        let footerFont = UIFont(name: "Lora-Italic", size: 11) ?? UIFont.italicSystemFont(ofSize: 11)
        let footerSize = (footer as NSString).size(withAttributes: [.font: footerFont])
        (footer as NSString).draw(
            at: CGPoint(x: (pageRect.width - footerSize.width) / 2, y: pageRect.height - margin - 24),
            withAttributes: [.font: footerFont, .foregroundColor: UIColor.gray]
        )
    }

    // MARK: - Helpers

    private static func coverColors(for coverStyleRaw: String) -> (UIColor, UIColor) {
        switch coverStyleRaw {
        case Story.CoverStyle.gradient.rawValue:
            return (UIColor(red: 0xE0/255, green: 0x78/255, blue: 0x56/255, alpha: 1),
                    UIColor(red: 0xC6/255, green: 0x5A/255, blue: 0x3E/255, alpha: 1))
        case Story.CoverStyle.typographic.rawValue:
            return (UIColor(red: 0xEA/255, green: 0xDF/255, blue: 0xCF/255, alpha: 1),
                    UIColor(red: 0xE8/255, green: 0xDC/255, blue: 0xC4/255, alpha: 1))
        default:
            return (UIColor(red: 0xE8/255, green: 0xDC/255, blue: 0xC4/255, alpha: 1),
                    UIColor(red: 0xF0/255, green: 0xE4/255, blue: 0xD7/255, alpha: 1))
        }
    }

    private static func monthYearStamp(_ iso: String) -> String? {
        let formatters: [ISO8601DateFormatter] = [
            { let f = ISO8601DateFormatter(); f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]; return f }(),
            { let f = ISO8601DateFormatter(); f.formatOptions = [.withInternetDateTime]; return f }(),
        ]
        for fmt in formatters {
            if let date = fmt.date(from: iso) {
                let df = DateFormatter()
                df.dateFormat = "MMM yyyy"
                return df.string(from: date).uppercased()
            }
        }
        return nil
    }

    private static func sanitizeFileName(_ name: String) -> String {
        let invalid = CharacterSet(charactersIn: "/\\:*?\"<>|")
        let cleaned = name.unicodeScalars
            .map { invalid.contains($0) ? "-" : String($0) }
            .joined()
            .trimmingCharacters(in: .whitespaces)
        return cleaned.isEmpty ? "Story" : cleaned
    }
}

File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/stories/StoryPdfExporter.kt
/*
 * StoryPdfExporter.kt
 *
 * Stories Phase 4 — renders a story + items into a US-Letter
 * `PdfDocument`. Mirror of iOS `StoryPdfExporter.swift`; see that
 * file's header for the per-page item count rule (full=1, half=2,
 * grid=4) + the layout choices.
 */

package app.quickink.mobile.features.stories

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Shader
import android.graphics.pdf.PdfDocument
import android.net.Uri
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import app.quickink.mobile.data.story.StoryEntity
import app.quickink.mobile.data.storyitem.StoryItemEntity
import java.io.File
import java.io.FileOutputStream

object StoryPdfExporter {

    private const val PAGE_WIDTH  = 612   // US Letter @ 72 DPI
    private const val PAGE_HEIGHT = 792
    private const val MARGIN      = 48f

    /**
     * Render the story to a temp PDF. Returns the on-disk file ready
     * for FileProvider sharing.
     *
     * `previewUris` maps each capture-backed `StoryItemEntity.refId`
     * to its on-disk `preview_uri`. The exporter loads + embeds the
     * bitmap when present; misses fall back to a cream-box.
     */
    fun export(
        context: Context,
        story: StoryEntity,
        items: List<StoryItemEntity>,
        previewUris: Map<String, String> = emptyMap(),
    ): File {
        val doc = PdfDocument()
        try {
            // Cover page.
            renderPage(doc, pageNum = 1) { canvas, _ ->
                drawCover(canvas, story, items)
            }

            // Item pages.
            val chunks = paginate(items)
            chunks.forEachIndexed { index, chunk ->
                renderPage(doc, pageNum = 2 + index) { canvas, _ ->
                    drawChunk(canvas, chunk, items, previewUris)
                }
            }

            // End-card page.
            renderPage(doc, pageNum = 2 + chunks.size) { canvas, _ ->
                drawEndCard(canvas)
            }

            val dir = File(context.cacheDir, "stories-exports").apply { if (!exists()) mkdirs() }
            val file = File(dir, "${sanitizeFileName(story.title)}.pdf")
            FileOutputStream(file).use { out -> doc.writeTo(out) }
            return file
        } finally {
            doc.close()
        }
    }

    private fun renderPage(doc: PdfDocument, pageNum: Int, draw: (Canvas, PdfDocument.Page) -> Unit) {
        val pageInfo = PdfDocument.PageInfo.Builder(PAGE_WIDTH, PAGE_HEIGHT, pageNum).create()
        val page = doc.startPage(pageInfo)
        draw(page.canvas, page)
        doc.finishPage(page)
    }

    // MARK: - Cover

    private fun drawCover(canvas: Canvas, story: StoryEntity, items: List<StoryItemEntity>) {
        val (startCol, endCol) = coverColors(story.coverStyle)
        val coverRect = RectF(MARGIN, 96f, PAGE_WIDTH - MARGIN, 96f + 320f)

        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            shader = LinearGradient(
                coverRect.left, coverRect.top, coverRect.right, coverRect.bottom,
                startCol, endCol, Shader.TileMode.CLAMP,
            )
        }
        canvas.drawRoundRect(coverRect, 16f, 16f, paint)

        val stampSource = story.timeRangeStart
            ?: items.firstOrNull()?.let { it.occurredAt ?: it.createdAt }
            ?: story.createdAt
        val stamp = monthYearStamp(stampSource) ?: ""
        val stampPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.DKGRAY
            textSize = 10f
            isFakeBoldText = false
            isAntiAlias = true
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            letterSpacing = 0.2f
        }
        canvas.drawText(stamp, coverRect.left + 16f, coverRect.top + 30f, stampPaint)

        val titlePaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.BLACK
            textSize = 26f
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.NORMAL)
        }
        val titleY = coverRect.bottom - 56f
        canvas.drawText(story.title, coverRect.left + 16f, titleY, titlePaint)

        if (!story.subtitle.isNullOrEmpty()) {
            val subPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                color = Color.DKGRAY
                textSize = 16f
                typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            }
            canvas.drawText(story.subtitle, coverRect.left + 16f, coverRect.bottom - 24f, subPaint)
        }

        val attribution = "${items.size} items"
        val attrPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.DKGRAY
            textSize = 11f
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            textAlign = Paint.Align.CENTER
        }
        canvas.drawText(attribution, PAGE_WIDTH / 2f, coverRect.bottom + 24f, attrPaint)
    }

    // MARK: - Chunks

    private data class Chunk(val items: List<StoryItemEntity>, val perPage: Int)

    private fun paginate(items: List<StoryItemEntity>): List<Chunk> {
        val out = mutableListOf<Chunk>()
        var i = 0
        while (i < items.size) {
            val perPage = itemsPerPage(items[i].layout)
            val end = (i + perPage).coerceAtMost(items.size)
            out += Chunk(items.subList(i, end), perPage)
            i = end
        }
        return out
    }

    private fun itemsPerPage(layoutRaw: String): Int = when (layoutRaw) {
        StoryItemEntity.Layout.HALF.raw -> 2
        StoryItemEntity.Layout.GRID.raw -> 4
        else                            -> 1
    }

    private fun drawChunk(canvas: Canvas, chunk: Chunk, allItems: List<StoryItemEntity>, previewUris: Map<String, String>) {
        val markers = StoryDayMarkers.derive(allItems).associateBy { it.precedingItemId }
        val contentTop = MARGIN
        val contentLeft = MARGIN
        val contentRight = PAGE_WIDTH - MARGIN
        val contentBottom = PAGE_HEIGHT - MARGIN
        var y = contentTop

        for (item in chunk.items) {
            markers[item.id]?.let { marker ->
                y = drawDayMarker(canvas, marker.label, y, contentLeft)
            }
            val height = when (chunk.perPage) {
                2 -> (contentBottom - contentTop - 32f) / 2f
                4 -> (contentBottom - contentTop - 32f * 2) / 4f
                else -> contentBottom - y
            }
            val rect = RectF(contentLeft, y, contentRight, y + height)
            drawItem(canvas, item, rect, previewUris)
            y = rect.bottom + 24f
        }
    }

    private fun drawDayMarker(canvas: Canvas, label: String, atY: Float, left: Float): Float {
        val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.parseColor("#C65A3E")
            textSize = 11f
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            letterSpacing = 0.1f
        }
        canvas.drawText(label, left, atY + 12f, paint)
        return atY + 24f
    }

    private fun drawItem(canvas: Canvas, item: StoryItemEntity, rect: RectF, previewUris: Map<String, String> = emptyMap()) {
        when (item.kind) {
            StoryItemEntity.Kind.TEXT_BLOCK.raw -> {
                val text = item.text.orEmpty()
                drawWrappedText(canvas, text, rect, 14f, italic = false, color = Color.BLACK)
            }
            StoryItemEntity.Kind.HANDWRITTEN_NOTE.raw -> {
                val rule = Paint().apply { color = Color.parseColor("#C65A3E") }
                canvas.drawRect(rect.left, rect.top, rect.left + 2f, rect.top + minOf(rect.height(), 60f), rule)
                drawWrappedText(canvas, item.text.orEmpty(), RectF(rect.left + 14f, rect.top, rect.right, rect.bottom), 18f, italic = true, color = Color.DKGRAY)
            }
            StoryItemEntity.Kind.DATE_DIVIDER.raw -> {
                val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                    color = Color.parseColor("#C65A3E")
                    textSize = 12f
                    typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                    letterSpacing = 0.08f
                    textAlign = Paint.Align.CENTER
                }
                canvas.drawText(item.text ?: "Date divider", (rect.left + rect.right) / 2f, rect.top + 16f, paint)
            }
            StoryItemEntity.Kind.PLACE_PIN.raw -> {
                val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                    color = Color.DKGRAY
                    textSize = 13f
                    typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                }
                canvas.drawText("📍 ${item.text.orEmpty()}", rect.left, rect.top + 16f, paint)
            }
            StoryItemEntity.Kind.VOICE_CLIP.raw -> {
                val stroke = Paint(Paint.ANTI_ALIAS_FLAG).apply {
                    style = Paint.Style.STROKE
                    strokeWidth = 1f
                    color = Color.LTGRAY
                }
                val inner = RectF(rect.left, rect.top + 8f, rect.right, rect.bottom - 8f)
                canvas.drawRoundRect(inner, 12f, 12f, stroke)
                val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                    color = Color.DKGRAY
                    textSize = 12f
                    typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                }
                canvas.drawText("🎙  ${item.caption ?: "Voice clip — open the app to listen."}", rect.left + 12f, rect.top + 28f, paint)
            }
            else -> {
                val photoHeight = minOf(rect.height() - 30f, 360f)
                val photoRect = RectF(rect.left, rect.top, rect.right, rect.top + photoHeight)
                val bmp = item.refId?.let { previewUris[it] }?.let { loadBitmap(it) }
                if (bmp != null) {
                    canvas.save()
                    val clip = Path().apply { addRoundRect(photoRect, 8f, 8f, Path.Direction.CW) }
                    canvas.clipPath(clip)
                    val dst = aspectFill(bmp.width, bmp.height, photoRect)
                    canvas.drawBitmap(bmp, null, dst, null)
                    canvas.restore()
                    bmp.recycle()
                } else {
                    val fill = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.parseColor("#EAE5DA") }
                    canvas.drawRoundRect(photoRect, 8f, 8f, fill)
                }
                if (!item.caption.isNullOrEmpty()) {
                    val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                        color = Color.DKGRAY
                        textSize = 12f
                        typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                    }
                    canvas.drawText(item.caption, rect.left, photoRect.bottom + 14f, paint)
                }
            }
        }
    }

    /** Load a bitmap from a `file://` URI when the file is present. */
    private fun loadBitmap(uri: String): Bitmap? = runCatching {
        val parsed = if (uri.startsWith("file://")) Uri.parse(uri) else null
        val path = parsed?.path ?: uri
        BitmapFactory.decodeFile(path)
    }.getOrNull()

    /** Aspect-fill rect for a bitmap of `(w, h)` into `target`. */
    private fun aspectFill(w: Int, h: Int, target: RectF): RectF {
        if (w <= 0 || h <= 0) return target
        val scale = maxOf(target.width() / w, target.height() / h)
        val outW = w * scale
        val outH = h * scale
        val cx = (target.left + target.right) / 2f
        val cy = (target.top + target.bottom) / 2f
        return RectF(cx - outW / 2f, cy - outH / 2f, cx + outW / 2f, cy + outH / 2f)
    }

    // MARK: - End-card

    private fun drawEndCard(canvas: Canvas) {
        val rect = RectF(MARGIN, 280f, PAGE_WIDTH - MARGIN, 460f)
        val stroke = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.STROKE
            strokeWidth = 1f
            color = Color.LTGRAY
        }
        canvas.drawRoundRect(rect, 14f, 14f, stroke)
        val end = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.DKGRAY
            textSize = 13f
            letterSpacing = 0.12f
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            textAlign = Paint.Align.CENTER
        }
        canvas.drawText("— THE END —", PAGE_WIDTH / 2f, rect.centerY(), end)

        val footer = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.GRAY
            textSize = 11f
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            textAlign = Paint.Align.CENTER
        }
        canvas.drawText("— Made with QuickInk · scan, jot, find again. —",
            PAGE_WIDTH / 2f, PAGE_HEIGHT - MARGIN - 16f, footer)
    }

    // MARK: - Helpers

    private fun drawWrappedText(canvas: Canvas, text: String, rect: RectF, size: Float, italic: Boolean, color: Int) {
        val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            this.color = color
            this.textSize = size
            typeface = android.graphics.Typeface.create(
                android.graphics.Typeface.SERIF,
                if (italic) android.graphics.Typeface.ITALIC else android.graphics.Typeface.NORMAL,
            )
        }
        val width = rect.width().toInt().coerceAtLeast(1)
        val layout = StaticLayout.Builder
            .obtain(text, 0, text.length, paint, width)
            .setAlignment(Layout.Alignment.ALIGN_NORMAL)
            .setIncludePad(false)
            .setLineSpacing(4f, 1f)
            .build()
        canvas.save()
        canvas.translate(rect.left, rect.top)
        layout.draw(canvas)
        canvas.restore()
    }

    private fun coverColors(coverStyleRaw: String): Pair<Int, Int> = when (coverStyleRaw) {
        StoryEntity.CoverStyle.GRADIENT.raw    -> Color.parseColor("#E07856") to Color.parseColor("#C65A3E")
        StoryEntity.CoverStyle.TYPOGRAPHIC.raw -> Color.parseColor("#EADFCF") to Color.parseColor("#E8DCC4")
        else                                   -> Color.parseColor("#E8DCC4") to Color.parseColor("#F0E4D7")
    }

    private fun monthYearStamp(iso: String): String? = runCatching {
        val dt = java.time.OffsetDateTime.parse(iso)
        java.time.format.DateTimeFormatter.ofPattern("MMM yyyy", java.util.Locale.ENGLISH).format(dt).uppercase()
    }.getOrNull()

    private fun sanitizeFileName(name: String): String {
        val out = name.map { c -> if (c in "/\\:*?\"<>|") '-' else c }.joinToString("").trim()
        return out.ifBlank { "Story" }
    }
}
